### PR TITLE
[CSL-1602] Improve update downloading logic

### PR DESCRIPTION
**Pull request by Dolgakov**
Improve downloadUpdateDo according to (1) [(https://github.com/input-output-hk/cardano-sl/blob/develop/docs/proposals/update-downloader.md#suggested-improvements)]

Instead of "Do not even start checkForUpdateWorker if the list of servers is empty (which is the case for core nodes)" server list checked for emptiness inside checkForUpdate with noServersProvided function. I made so because It seems that it`s better to check servers of one relevant ConfirmedProposalState (newestCPS) than all NonEmpty ConfirmedProposalState.

### CODE REVIEW

File: update/Pos/Update/Download.hs [hunk 1]
```diff
@@ -129,18 +129,35 @@ downloadUpdateDo updHash cps@ConfirmedProposalState {..} = do
...
+                file <- ExceptT $ downloadHash updateServers updHash <&>
+                        first (sformat ("Update download (hash "%build%
+                                        ") has failed: "%stext) updHash)
+        
+                logInfo $ "Update was downloaded, saving to " <> show updPath
+        
+                liftIO $ BSL.writeFile updPath file
+                logInfo $ "Update was downloaded, saved to " <> show updPath
+
+        fileIsOnDrive <- liftIO $ (doesFileExist updPath)
```
Comment (hithroc) [inline update/Pos/Update/Download.hs]: Either reduntant parenthesis or `$`

File: update/Pos/Update/Download.hs [hunk 2]
```diff
@@ -129,18 +129,35 @@ downloadUpdateDo updHash cps@ConfirmedProposalState {..} = do
...
+        if fileIsOnDrive
+            then do
+                logInfo $ sformat ("Found an update file: "%stext) (toText updPath)
+                localUpdateFile <- liftIO $ BSL.readFile updPath -- could be an error while reading
+                validPkgHash <- lift $ getUpdateHash cps
+                let fileIsValidUpdate = (installerHash localUpdateFile) == validPkgHash
```
Comment (hithroc) [inline update/Pos/Update/Download.hs]: Is there really a need for this variable?

File: update/Pos/Update/Download.hs [hunk 3]
```diff
@@ -210,6 +227,6 @@ downloadUri manager uri h = do
 * check timeouts?
 * how should we in general deal with e.g. 1B/s download speed?
 * if we expect updates to be big, use laziness/conduits (httpLBS isn't lazy,
-  despite the “L” in its name)
+despite the “L” in its name)
```
Comment (hithroc) [inline update/Pos/Update/Download.hs]: Why remove formatting?

File: update/Pos/Update/Worker.hs [hunk 4]
```diff
@@ -12,6 +12,9 @@ import           Formatting (build, sformat, (%))
 import           Serokell.Util.Text (listJsonIndent)
 import           System.Wlog (logDebug, logInfo)
 
+import           Control.Lens (views)
+
+
```
Comment (hithroc) [inline update/Pos/Update/Worker.hs]: Two blank lines is a bit too much for separation, and `Control.Lens` import should be part of the external imports chunk above, not separate from it.

File: update/Pos/Update/Worker.hs [hunk 5]
```diff
@@ -23,6 +26,7 @@ import           Pos.Update.Configuration (curSoftwareVersion)
 import           Pos.Update.Context (UpdateContext (..))
 import           Pos.Update.DB (getConfirmedProposals)
 import           Pos.Update.Download (downloadUpdate)
+import           Pos.Update.Params (UpdateParams (..))
```
Comment (hithroc) [inline update/Pos/Update/Worker.hs]: Not an alphabetically sorted import.

**Comment by hithroc** [comment]
Also I've noticed that this PR and #2612 have exactly the same changes to Worker.hs while having different commits.

**Comment by gromakovsky** [comment]
I'd like to review this PR after @hithroc approves it.

File: update/Pos/Update/Worker.hs [hunk 6]
```diff
@@ -67,14 +69,21 @@ checkForUpdate = do
...
+                    ("There are new confirmed update proposals for our application: "
+                    %listJsonIndent 2%
+                    "\n The newest one is: "%build%" and we want to download it")
+                    (cpsUpdateProposal <$> confirmedProposals)
+                    (cpsUpdateProposal newestCPS)
+            downloadUpdate newestCPS
+
+    noServersProvided :: ConfirmedProposalState -> m Bool
+    noServersProvided ConfirmedProposalState {..} = 
+        null <$> views (lensOf @UpdateParams) upUpdateServers
```
Comment (gromakovsky) [inline update/Pos/Update/Worker.hs]: I do not understand this. `ConfirmedProposalState` is not used by this function. Why is it passed here?
I wrote it earlier in #2612, btw.

File: update/Pos/Update/Download.hs [hunk 7]
```diff
@@ -129,18 +129,36 @@ downloadUpdateDo updHash cps@ConfirmedProposalState {..} = do
...
+                logInfo $ "Update was downloaded, saved to " <> show updPath
+
+        -- Check if an update is already downloaded to `updPath`.
+        -- If it is valid one, it would be passed to `downloadedMVar`.
+        fileIsOnDrive <- liftIO $ doesFileExist updPath
+        if fileIsOnDrive
+            then do
+                logInfo $ sformat ("Found an update file: "%stext) (toText updPath)
+                localUpdateFile <- liftIO $ BSL.readFile updPath -- could be an error while reading
+                validPkgHash <- lift $ getUpdateHash cps
```
Comment (gromakovsky) [inline update/Pos/Update/Download.hs]: This is `updHash`, it's passed to this function, don't calculate it twice.

File: update/Pos/Update/Download.hs [hunk 8]
```diff
@@ -129,18 +129,36 @@ downloadUpdateDo updHash cps@ConfirmedProposalState {..} = do
...
+
+                if (installerHash localUpdateFile) == validPkgHash
```
Comment (gromakovsky) [inline update/Pos/Update/Download.hs]: Redundant brackets.

**Comment by gromakovsky** [comment]
> I'd like to review this PR after @hithroc approves it.

Next time please ping me more aggressively before merging.

File: update/Pos/Update/Worker.hs (hunk 6 above)
Comment (Dolgakov) [reply update/Pos/Update/Worker.hs]: Sorry, I just have forgotten about that :(

**Pull request merged**